plugins/generator-1.18.2/forge-1.18.2/templates/elementinits/villagerprofessions.java.ftl

```diff
@@ -0,0 +1,84 @@
+ # the Free Software Foundation, either version 3 of the License, or
+ # (at your option) any later version.
+ #
+ # This program is distributed in the hope that it will be useful,
+ # but WITHOUT ANY WARRANTY; without even the implied warranty of
+ # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ # GNU General Public License for more details.
+ #
+ # You should have received a copy of the GNU General Public License
+ # along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ #
+ # Additional permission for code generator templates (*.ftl files)
+ #
+ # As a special exception, you may create a larger work that contains part or
+ # all of the MCreator code generator templates (*.ftl files) and distribute
+ # that work under terms of your choice, so long as that work isn't itself a
+ # template for code generation. Alternatively, if you modify or redistribute
+ # the template itself, you may (at your option) remove this special exception,
+ # which will cause the template and the resulting code generator output files
+ # to be licensed under the GNU General Public License without this special
+ # exception.
+-->
+
+<#-- @formatter:off -->
+
+<#include "../mcitems.ftl">
+
+/*
+*    MCreator note: This file will be REGENERATED on each build.
+*/
+
+package ${package}.init;
+
+import net.minecraft.sounds.SoundEvent;
+
+public class ${JavaModName}VillagerProfessions {
+
+    public static final DeferredRegister<PoiType> POI = DeferredRegister.create(ForgeRegistries.POI_TYPES, ${JavaModName}.MODID);
+    public static final DeferredRegister<VillagerProfession> PROFESSIONS = DeferredRegister.create(ForgeRegistries.PROFESSIONS, ${JavaModName}.MODID);
+
+    <#list villagerprofessions as villagerprofession>
+        public static final RegistryObject<VillagerProfession> ${villagerprofession.getModElement().getRegistryNameUpper()} =
+            registerProfession(
+			    "${villagerprofession.getModElement().getRegistryName()}",
+                ${mappedBlockToBlock(villagerprofession.pointOfInterest)},
+                () -> new SoundEvent(new ResourceLocation("${villagerprofession.actionSound}"))
+            );
+    </#list>
+
+	private static RegistryObject<VillagerProfession> registerProfession(String name, Block block, Supplier<SoundEvent> soundEventSupplier) {
+		Optional<PoiType> existingCheck = PoiType.forState(block.defaultBlockState());
+
+		if (existingCheck.isPresent()) {
+            ${JavaModName}.LOGGER.error("Skipping villager profession " + name + " that uses POI block " + block + " that is already in use by " + existingCheck);
+			return null;
+		}
+
+		Supplier<PoiType> poi = POI.register(name, () -> new PoiType(name, ImmutableSet.copyOf(block.getStateDefinition().getPossibleStates()), 1, 1));
+        return PROFESSIONS.register(name, () -> new RegistrySafeVillagerProfession(${JavaModName}.MODID + ":" + name, poi.get(), ImmutableSet.of(), ImmutableSet.of(), soundEventSupplier));
```

Comment:
Are these sets taken away from the custom profession constructor just in case this functionality is added in the future?

Reply:
Done!

Reply:
@Defeatomizer you probably meant to fully inline those, not extend to one level higher? Or is the fix the one you meant?

Reply:
Yes, I meant passing these to the parent class directly from the subclass constructor, so this is not resolved yet.

Reply:
I don't really get what you mean

Reply:
The `ImmutableSet.of()` arguments should be passed from inside `RegistrySafeVillagerProfession` class constructor to its parent class instead of taking them all the way through registration method call and constructor parameters.

Reply:
I still don't get it. You talked about adding a functionality in the future but 1. The 1.18.2 version will be dropped by then (probably) and is the only version with a custom class, and 2. Inlining it would probably remove the way to add a functionality in the future, not make it better?

Reply:
The values of `specificItems` and `relatedWorldBlocks` used in `RegistrySafeVillagerProfession()` should not be arguments. I don't know how to explain that better without suggestions.
>You talked about adding a functionality in the future

I was just asking, not recommending anything.

Reply:
It's done. I removed the sets from the arguments